### PR TITLE
add force transfer did interface

### PR DESCRIPTION
**Pull request by ruibinx**
This pr adds an extrinsic which allows root user forcing transfer did to another account, and transfer all of its assets. It should be used for account recovery only.

**Type of change**

- [x] New Feature (non-breaking change which adds functionality)

### CODE REVIEW

File: pallets/traits/src/nfts.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+use sp_runtime::DispatchError;
+
+pub trait Nfts<AccountId> {
+    // force transfer all assets of account src to account dest
+    fn force_transfer_all_assets(src: &AccountId, dest: &AccountId) -> Result<(), DispatchError>;
+}
+
+impl<AccountId> Nfts<AccountId> for () {
+    fn force_transfer_all_assets(_src: &AccountId, _dest: &AccountId) -> Result<(), DispatchError> {
+        Ok(())
```
[the remark this answers is not in the document]
Comment (ruibinx) [reply pallets/traits/src/nfts.rs:10]: It's for unit tests. Mocks could use unit type as trait place holder if it doesn't test NFT.

File: pallets/did/src/lib.rs [hunk 2]
```diff
@@ -159,6 +164,34 @@ pub mod pallet {
             Ok(())
         }
 
+        #[pallet::weight(T::WeightInfo::transfer())]
+        pub fn force_transfer(
```
[the remark this answers is not in the document]
Comment (ruibinx) [reply pallets/did/src/lib.rs:168]: uniques指ad3吗？

File: pallets/did/src/lib.rs [hunk 3]
```diff
@@ -159,6 +164,34 @@ pub mod pallet {
...
+            ensure!(!<DidOf<T>>::contains_key(&dest), Error::<T>::DidExists);
+            let mut meta = <Metadata<T>>::get(did).ok_or(Error::<T>::DidNotExists)?;
+
+            let source = meta.account.clone();
+            meta.account = dest.clone();
+            meta.created = <frame_system::Pallet<T>>::block_number();
+
+            let ad3_balance = T::Currency::total_balance(&source);
+            T::Currency::transfer(&source, &dest, ad3_balance, AllowDeath)?;
+            T::Nfts::force_transfer_all_assets(&source, &dest)?;
```
[the remark this answers is not in the document]
Comment (ruibinx) [reply pallets/did/src/lib.rs:184]: > 建议：与其转nft_fragments，不如把所有的asset转过去，比如以后可能会出现非fragment类型的asset

开始的时候思路是这样的，问题是pallet-assets没有提供枚举asset的接口，存储也是private的

[the remark this answers is not in the document]
Comment (ruibinx) [reply pallets/did/src/lib.rs:184]: > force_transfer_all_nft_fragments

force_transfer_all_fractions? NFT trait下是不是可以不写nft了？

File: pallets/did/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (coderfengyun) [reply pallets/did/src/lib.rs:168]: unique是指nft，在substrate里nft叫unique

File: pallets/did/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (coderfengyun) [reply pallets/did/src/lib.rs:184]: 看起来OK啊

**Review by coderfengyun — APPROVED** [review]

File: pallets/did/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ruibinx) [reply pallets/did/src/lib.rs]: 改成force_transfer_with_assets了，概念模糊一点，之后可以继续扩展

**Pull request merged**